File: src/sci/impl/java/options/namespace.clj
(ns sci.impl.java.options.namespace)

(set! *warn-on-reflection* true)

(gen-class
 :name borkdude.sci.options.Namespace
 :constructors {[String] []}
 :state "state"
 :init "init"
 :prefix "opts-ns-")

(gen-class
 :name borkdude.sci.options.Namespace
 :constructors {[String] []}
 :state "state"
 :init "init"
 :prefix "opts-ns-"
 :methods [["addVar" [java.lang.String java.lang.Object] borkdude.sci.options.Namespace]
           ["val" [] java.lang.Object]])

(defn opts-ns-init [name]
  [[] (atom {:name (symbol name)})])

(defn opts-ns-addVar [^borkdude.sci.options.Namespace this k v]
  (let [state (.state this)]
    (swap! state assoc-in [:vars (symbol k)] v)
    this))

(defn opts-ns-val [^borkdude.sci.options.Namespace this]
  @(.state this))

File: src-java/borkdude/sci/options/Namespace.java
package borkdude.sci.options;

import java.util.HashMap;

public class Namespace {

    private String _name;
    private HashMap<String,Object> _vars = new HashMap<String, Object>();

    public Namespace(String name) {
        _name = name;
    }

    public String getName() {
        return _name;
    }

    public Namespace addVar(String name, Object value) {
        _vars.put(name, value);
        return this;
    }

    /**
     * Implementation detail, don't use.
     * @return: irrelevant.
     */
    public HashMap<String, Object> _val() {
        return _vars;
    }
}
